plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java

```diff
@@ -13,28 +13,111 @@
  */
 package org.apache.kudu.client;
 
+import io.trino.plugin.kudu.KuduClientSession;
+import io.trino.plugin.kudu.KuduClientWrapper;
 import io.trino.spi.TrinoException;
 
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkState;
 import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
 import static java.lang.String.format;
+import static org.apache.kudu.client.SessionConfiguration.FlushMode.MANUAL_FLUSH;
 
 /**
- * Operation.getChangeType() is package private
+ * This class is used to buffer operations and apply them in a batch and verify the batch was successfully applied.
+ * This gives:
+ * - performance of not flushing after every operation
+ * - correctness of not flushing in the background and hoping operations succeed
+ * The buffer is flushed when:
+ * - The KuduOperationApplier is closed
+ * - Or if the KuduOperationApplier reaches the bufferMaxOperations
+ * Note: Operation.getChangeType() is package private
  */
 public final class KuduOperationApplier
+        implements AutoCloseable
 {
-    private KuduOperationApplier()
+    private static final int bufferMaxOperations = 1000;
+
+    private int currentOperationsInBuffer;
```

Comment:
Should this be an Atomic or should we synchronize on this?

Reply:
Each KuduOperationApplier should only be used by a single thread, I can use `@NotThreadSafe` if we want to make that explicit

Reply:
cc: @findepi Do we have convention? I see some classes marked `@ThreadSafe` though. Do we assume not thread safe if not annotated?

Reply:
> Do we assume not thread safe if not annotated?

more or less (ie unless it's clear the class needs to be thread safe

we don't have a convention to annotate `@NotThreadSafe` those who're not.

Reply:
I'll remove the annotation and add to the javadoc then 👍